rfcs/RFC-Uploads.md

```diff
@@ -0,0 +1,377 @@
+
+#### Cloud to Cloud Upload Flow
+
+Cloud to cloud upload requires the user to provide a shareable link for a file on their CSP from which DP can upload
+the file to the S3 bucket. Now to walk through the cloud to cloud flow, see figure 1.
+
+![Cloud to Cloud](https://app.lucidchart.com/publicSegments/view/e1e72143-c108-4c7f-bf6c-053d98256c1f/image.png)
+**Figure 2:** Architecture of a user uploading from their cloud to our cloud.
+
+##### 1. Get Share Link
+
+The user must generate a sharable link from their CSP. Both Google Drive and Dropbox support the generation of
+sharable links. Only sharable links should only point to a single file, not a folder or submission of files.
+
+##### 2. Share Link
+
+Using the browser App, make a \_POST {submission_id}/upload/link_request to upload the file
+by providing the sharable link in the request body. The response will contain the dataset_id, that can be used to
+check the status of the upload
+
+##### 3. Sanitize
+
+To prevent [server side request forgery](https://owasp.org/www-community/attacks/Server_Side_Request_Forgery) the
+incoming shared link will be sanitized before queuing for download. The links generated by CSP are often not direct
+links to the files and often need to be parsed into the correct shape for downloading. The correct URL will be generated
+as part of the sanitization process. The backend that runs this should also be given minimal privileges to further protect from
+server-side request forgery.
+
+###### How to download from a shared link
+
+- Dropbox <https://help.dropbox.com/files-folders/share/force-download>
+- Google Drive <https://www.labnol.org/internet/direct-links-for-google-drive/28356/>
+
+##### 4. _Optional_ Queue
+
+This part is optional and depends on the load expected and server type selected. The queue stores the links that need to
+be processed. It buffers the downstream servers from download requests. This allows us to spin up servers as needed.
+It also gives us the ability to retry a download if one of the servers fails to download or upload the file.
+
+###### Why might the download fail
+
+- The downloading server ran out of storage.
+- The downloading server crashed.
+- The CSP experience and outage
+- The user has reached their daily bandwidth allowance from their CSP. [more info](https://help.dropbox.com/files-folders/share/banned-links)
+
+##### 5. Start EC2 Instance to Download File
+
+Spin up EC2 instances with enough storage to download the whole file. TODO: See if spot instances could be used.
+
+##### 6. Download the File
+
+The most time will be spent here downloading the file. Most CSPs offer a way to validate the integrity of your download
+by providing a hash of the data. This hash can be compared with the hash of the downloaded file. If the hash does not match,
+then reattempt the download.
+
+##### 7. Upload to S3
+
+If the file is >5GB in size, upload the file directly to s3 without writing to disk. For files <5GB, the file must
```

Comment:
Actually, I've found out that this can be done without downloading the file completely. Initially, when I read how s3 multipart upload works, it said the file needed to be broken up before it could be uploaded. I misinterpreted this as broken up all at once. Now I see that the s3 client will handle the breaking up for you, and as long as you keep streaming in chunks it will handle the rest of the upload process.

This does have the drawback of not verifying the integrity of the download from the third-party before uploading to our cloud. This goes for all files. We would need to find a way to calculate the 3rd party hash as we go which may not be possible depending on what hash algorithm they use.

Reply:
I'll modify the spec to reflect this.